cryptpad: 3.20.1 -> 4.9.0

DavHau opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
see https://github.com/NixOS/nixpkgs/issues/128154

What worries me a bit is that it requires bootstrap 3.1.1 and jquery 2.1.x for some reason which are EOL.
Not sure if this is due to the tooling used here to update the dependencies or if those are actual upstream requirements. I need to investigate this further.


###### Things done
- pump version of cryptpad
- updated `generate.sh` to use a working version of bower2nix and execute bower2nix automatically
- added missing bower dependencies via override (bootstrap 3.1.1, jquery 2.1.0)
- update nodePackages
- update bower-packages for cryptpad

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-21requirements11-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

DavHau committed:
cryptpad: 3.20.1 -> 4.9.0

happysalada commented on pkgs/servers/web-apps/cryptpad/generate.sh:
```diff
@@ -1,11 +1,16 @@
 #!/usr/bin/env nix-shell
-#! nix-shell -i bash  -I nixpkgs=../../../.. -p nodePackages.node2nix nodePackages.bower2nix
+#! nix-shell -i bash  -I nixpkgs=../../../.. -p nodePackages.node2nix -p nix
```
you might not need the second `-p` here

happysalada commented on pkgs/servers/web-apps/cryptpad/generate.sh:
```diff
@@ -1,11 +1,16 @@
+cryptpadSrc=$(nix eval '(import ../../../.. {}).cryptpad.src' --raw)
+echo "cryptpad src: $cryptpadSrc"
+
+nix-shell -I nixpkgs=https://github.com/nixos/nixpkgs/tarball/18.03 -p nodePackages.bower2nix --run "bower2nix $cryptpadSrc/bower.json bower-packages.nix"
```
I'm wondering a little how long that will be available. Maybe forever since this is just a git branch, but I'm interested in knowing for sure.

SuperSandro2000 replied:
Do we really need to use a **3** year old tarball?!

happysalada replied:
From nixpkgs dev channel, those will be available for ever, just don't receive any updates.

happysalada replied:
Potentially node-10 is working and available on master. Dave just said he is going to have a look.

DavHau replied:
now using nodejs-10 instead of pulling old nixpkgs

SuperSandro2000 replied:
The problem is not that they might become unavailable but that you probably need to pull multiple gigabytes to build nodejs 10 in there.

happysalada replied:
Thank you for the info!

happysalada commented:
While building I noticed the following
```
/nix/store/37cqg4vgwykndi8b2sv6b0wis18mcr20-cryptpad-4.9.0/lib/node_modules/cryptpad/node_modules/mime/cli.js: interpreter directive changed from "#!/usr/bin/env node" to "/nix/store/qmv0vl6qk7j51pw10dghn91k0liddcki-nodejs-14.17.4/bin/node"
```

this is using the default nodejs, we have to check if there isn't a way to override to use node12 (which is what they use according to their readme).
https://docs.cryptpad.fr/en/dev_guide/setup.html

happysalada commented:
At the top level in all-packages it should be 
```
  cryptpad = callPackage ../servers/web-apps/cryptpad {
    nodejs = nodejs-12_x;
  };
```

happysalada commented:
I've got some small formatting differences.
Would you mind running the `nixpkgs-fmt` ?

happysalada commented:
I've checked the module it looks like nothing has changed, so it looks like the service module is still valid.

happysalada commented:
two more minor comments
Perhaps add in the generate.sh the two following things
- `This script is meant to be run in the current directory`. You can also play with `$0` if you like, but I don't think it's worth it here.
- `This script will not run on `darwin`. I get the following error, it seems to be due to something in nixpkgs 18.03
```
error: builder for '/nix/store/78fa13id79zx0kc1dcj6fqvqr5676lw2-pinpointDependencies.js.drv' failed due to signal 6 (Abort trap: 6);
       last 3 log lines:
       > dyld: Library not loaded: /usr/lib/system/libsystem_network.dylib
       >   Referenced from: /nix/store/46qj01lcjz48wb4hc7mvn3mhb3syda7h-Libsystem-osx-10.11.6/lib/libSystem.B.dylib
       >   Reason: image not found
       For full logs, run 'nix log /nix/store/78fa13id79zx0kc1dcj6fqvqr5676lw2-pinpointDependencies.js.drv'.
error: 1 dependencies of derivation '/nix/store/vipcl302bzd4g1ml9xrgyqx1jnspki26-node-bower2nix-3.2.0.drv' failed to build
error: error: build of '/nix/store/sx770ma2l7wj45prnyfbn04snw6zf0wm-stdenv-darwin.drv', '/nix/store/vipcl
```

happysalada commented:
Last but not least, do you have a nixos server ? Did you try to run this ?

happysalada commented:
on the nixkgs dev somebody mentioned that bower2nix also works with node-10
this is their overlay
```
bower2nix = super.nodePackages.bower2nix.overrideAttrs (oA: {
    nodejs = super.nodejs-10_x;
    buildInputs = [ super.nodejs-10_x ] ++ oA.buildInputs;
  });
```
do you think you could try with nodejs-10 ? (we should be able to find a more recent version of nixpkgs that has it)

DavHau committed:
cryptpad: generate.sh: stop using outdated nixpkgs

DavHau committed:
cryptpad: use nodejs12

DavHau commented:
Added some commits:
 - Removed script dependency on old nixpkgs (maybe this fixes the script for darwin as well)
 - use nodejs12 for cryptpad

> I've got some small formatting differences.
> Would you mind running the `nixpkgs-fmt` ?

AFAIK there is no community consensus on formatting. And as long as this is the case I see no point in spending time on subtle differences. It's not even clear which formatting tool is the `correct` one. If I'm wrong, please let me know.

> Last but not least, do you have a nixos server ? Did you try to run this ?

I did not yet test the module. That still needs to be done before this can be merged.

> on the nixkgs dev somebody mentioned that bower2nix also works with node-10

I did that now and therefore got rid of the outdated nixpkgs in the script. Thanks a lot for the hint.

happysalada commented:
On the formatting.
`nixpkgs-fmt` was a tool that was made specifically for this. But yes, it's not official that this will be the tool chosen. I feel it's better to choose one formatter. Feel free to skip if you feel like it.

So I believe what is left is just checking where those override of insecure bootstrap and jquery come from, correct?

DavHau committed:
cryptpad: add test for nixos module

happysalada commented:
Great adding a test!

DavHau committed:
cryptpad: reformat with nixpkgs-fmt

DavHau commented:
Re-formatted with nixpkgs-fmt.

About the EOL dependencies: I found out that the root of this issue is a conflict in dependency resolution. I tried using bower in a node12 docker image:
```
Please note that,
    bootstrap-tokenfield#0.12.1 depends on bootstrap#~3.1.1 which resolved to bootstrap#3.1.1
    cryptpad depends on bootstrap#^v4.0.0 which resolved to bootstrap#4.6.0
Resort to using bootstrap#^v4.0.0 which resolved to bootstrap#4.6.0
Code incompatibilities may occur.
```

One can see here that bower ignores the older version of bootstrap and instead installs the newer one, despite both versions are required.

I verified that also `bower install --offline` in `buildBowerComponents` installs the new version and does not install the outdated one.

The only reason why we need to also fetch the outdated bootstrap and jquery is that `bower install --offline` crashes otherwise. But it does not install them.

From my side everything is ready now.

DavHau committed:
cryptpad: add comment about EOL dependencies

happysalada commented:
There was a problem on github. Let me try to trigger ofborg tomorrow just to be safe

cole-h commented:
@ofborg eval

happysalada commented:
Result of `nixpkgs-review pr 133202` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cryptpad</li>
  </ul>
</details>

happysalada commented:
The build succeeds and it at looks good, thank you!
